docs/proposals/provider-agnostic-dns-health-checks.md

```diff
@@ -0,0 +1,77 @@
+# Provider agnostic DNS Health checks
+
+## Introduction
+
+The MGC has the ability of extending the DNS configuration of the gateway with the DNSPolicy resource. This resource allows users to configure health checks. As a result of configuring health checks, the controller creates the health checks in Route53, attaching them to the related DNS records. This has the benefit of automatically disabling an endpoint if it becomes unhealthy, and enabling it again when it becomes healthy again.
+This feature has a few shortfalls:
+
+1. It’s tightly coupled with Route53. If other DNS providers are supported they must either provide a similar feature, or health checks will not be supported
+2. Lacks the ability to reach endpoints in private networks
+
+This document describes a proposal to extend the current health check implementation to overcome these shortfalls.
+
+### Goals
+
+* Ability to configure health checks in the DNSPolicy associated to a Gateway
+* Ability for the health checks to reach endpoints in private networks
+* DNS records are disabled when the associated health check fails
+* Current status of the defined health checks is visible to the end user
+* ~~Transparently keep support for other health check providers like Route53~~
+
+## Proposal
+
+#### `DNSPolicy` resource
+
+The presence of the `healthCheck` field in the DNSPolicy will affect the reconciliation
+of the endpoints when creating/updating DNSRecords:
+
+* For every endpoint that is generated, a health check is created based on the configuration in the DNSPolicy
+* Relevant `providerSpecific` fields are added to the endpoint:
+  * A reference to the equivalent `DNSHealthCheckProbe` in the downstream cluster
+  will be set: `health-probe/name`, `health-probe/namespace`
+  * The health status of the `DNSHealthCheckProbe` will be propagated
```

Comment:
If you setup a relationship of a `DNSRecord` owns `DNSHealthCheckProbes` that the dnspolicy is creating, do you need to set this `providerSpecific` field? 

In order to keep the DNSPolicy reconciler clean it would be better if we did not have to add these values. Doing this, you either have to do what we currently do and pull all the records again after having just created/updated them, or combine the creation/deletion of `DNSHealthCheckProbes` resources into the `dnspolicy_dnsrecord` reconcile.

Would it be a better solution to just lookup healthchecks owned by the current dnsrecord during the dnsrecord reconciliation and build up a list of hosts/ips and their current health status and pass that into the [MultiClusterGatewayTarget](https://github.com/Kuadrant/multicluster-gateway-controller/blob/main/pkg/controllers/dnspolicy/dnspolicy_dnsrecords.go#L111), which will ultimately use that when deciding the weight for the record with that host/ip https://github.com/Kuadrant/multicluster-gateway-controller/blob/main/pkg/dns/service.go#L234